include/vast/ABI/Classify.hpp

```diff
@@ -232,6 +237,83 @@ namespace vast::abi
             return mlir::IntegerType::get( t.getContext(), final_size );
         }
 
+        static auto get_is_sse_type() {
+            return [](auto t) {
+                return is_scalar_float(t);
+            };
+        }
+
+        // TODO: In `clang` int and sse variants have separate implementations (there is
+        //       a lot of difference) but in the core they are doing very similar things.
+        //       It should be possible to make a shared helper that takes away a lot of
+        //       shared complexity.
+        static mlir_type sse_type_at_offset(mlir_type t, std::size_t offset,
+                                            mlir_type root, std::size_t root_offset,
+                                            const auto &ctx ) {
+            const auto &[ dl, _ ] = ctx;
+            auto &mctx = *t.getContext();
+
+            auto t0 = type_at_offset(t, offset, ctx, get_is_sse_type());
+            if (!t0 || size(dl, t0) == 64)
+                return fN(mctx, 64);
+
+            auto source_size = size(dl, root);
+            auto t0_size = size(dl, t0) - root_offset;
+
+            auto is_16b_fp = [&, dl=dl](auto type) {
+                return is_scalar_float(type) && size(dl, type) == 16;
+            };
+
+            // Fetch second type if applicable.
+            auto t1 = [&]() -> mlir_type {
+                auto nested = type_at_offset(t, offset + t0_size, ctx, get_is_sse_type());
+                if (t0_size > source_size && nested)
+                    return nested;
+
+                // We know `nested` was either not retrieved or that the sizes are not
+                // good enough.
+
+                // Check case of half/bfloat + float.
+                if (size(dl, t0) == 16 && source_size > 4) {
+                    // `+4` comes from alignement
```

Comment:
is this alignement fixed? if it can be altered from outside make it defensive to catch such cases.

Reply:
It is hardcoded in clang as well so I doubt it can be changed as this is result of how underlying arch works (other archs will do this completely differently).